gateway/src/apicast/mapping_rule.lua

```diff
@@ -27,7 +27,9 @@ local function hash_to_array(hash)
 end
 
 local function regexpify(pattern)
-  return pattern:gsub('?.*', ''):gsub("{.-}", '([\\w_.-]+)'):gsub("%.", "\\.")
+  -- as per the RFC: https://tools.ietf.org/html/rfc3986#section-3.3
+  local wildcard_regex = [[([\w-.~%%!$&'()*+,;=@:]+)]] -- using long Lua brackets [[...]], escaping `%`
+  return pattern:gsub('?.*', ''):gsub("{.-}", wildcard_regex):gsub("%.", "\\.")
```

Comment:
I'd like to try https://github.com/3scale/apicast/pull/707 on this to see if we can improve performance of it. The `:gsub` should not be JITable, so replacing it could help significantly in micro benchmarks.

Reply:
```lua
require('benchmark.ips')(function(b)
  b.time = 5
  b.warmup = 2

  local gsub = string.gsub
  local re_gsub = ngx.re.gsub

  local transforms = {
    { '?.*', '' },
    { "{.-}", [[([\w-.~%%!$&'()*+,;=@:]+)]] },
    { "%.", [[\.]] },
  }

  local function loop(pattern)
    for i=1, #transforms do
      pattern = gsub(pattern, transforms[i][1], transforms[i][2])
    end
    return pattern
  end

  local function closure(pattern)
    pattern = gsub(pattern, '?.*', '')
    pattern = gsub(pattern, "{.-}", [[([\w-.~%%!$&'()*+,;=@:]+)]])
    pattern = gsub(pattern, "%.", "\\.")

    return pattern
  end

  local function global(pattern)
    -- as per the RFC: https://tools.ietf.org/html/rfc3986#section-3.3
    local wildcard_regex = [[([\w-.~%%!$&'()*+,;=@:]+)]] -- using long Lua brackets [[...]], escaping `%`
    return pattern:gsub('?.*', ''):gsub("{.-}", wildcard_regex):gsub("%.", "\\.")
  end

  local function pcre(pattern)
    pattern = re_gsub(pattern, [[\?.*]], '')
    pattern = re_gsub(pattern, [[\{.+?\}]], [[([\w-.~%!$$&'()*+,;=@:]+)]])
    pattern = re_gsub(pattern, [[\.]], [[\.]])

    return pattern
  end

  local function pcre_jit(pattern)
    pattern = re_gsub(pattern, [[\?.*]], '', 'oj')
    pattern = re_gsub(pattern, [[\{.+?\}]], [[([\w-.~%!$$&'()*+,;=@:]+)]], 'oj')
    pattern = re_gsub(pattern, [[\.]], [[\.]], 'oj')

    return pattern
  end

  local str = '/tfgoodosod/{foo}/sdsd?asdasdd'

  print('loop:   ' , loop(str))
  print('closure:' , closure(str))
  print('global: ' , global(str))
  print('pcre:   ' , pcre(str))

  b:report('loop', function() return loop(str) end)
  b:report('gsub', function() return closure(str) end)
  b:report('string.gsub', function() return global(str) end)
  b:report('re.gsub', function() return pcre(str) end)
  b:report('re.gsub jit', function() return pcre_jit(str) end)

  b:compare()
end)
```

```
Warming up --------------------------------------
                loop
     68352 i/100ms

                gsub
     69712 i/100ms

         string.gsub
     38264 i/100ms

             re.gsub
       242 i/100ms

         re.gsub jit
       807 i/100ms

Calculating -------------------------------------
   730569.9 (±0.9%) i/s -    3691008 in   5.052678s
   739498.1 (±1.3%) i/s -    3764448 in   5.091396s
   731344.9 (±2.0%) i/s -    3673344 in   5.024815s
      851.0 (±18.4%) i/s -       4114 in   5.002932s
     4226.9 (±3.4%) i/s -      21789 in   5.161376s
```

So, could you use `gsub` function from the closure to get 2x performance ? :)

Reply:
Updated results using `resty.core` and JITable PCRE:

<details><summary> Benchmark code </summary><p>

```lua
local re_gsub = ngx.re.gsub

require('resty.core')
require('benchmark.ips')(function(b)
  b.time = 5
  b.warmup = 2

  local gsub = string.gsub

  local ffi_re_gsub = ngx.re.gsub

  local transforms = {
    { '?.*', '' },
    { "{.-}", [[([\w-.~%%!$&'()*+,;=@:]+)]] },
    { "%.", [[\.]] },
  }

  local function loop(pattern)
    for i=1, #transforms do
      pattern = gsub(pattern, transforms[i][1], transforms[i][2])
    end
    return pattern
  end

  local function closure(pattern)
    pattern = gsub(pattern, '?.*', '')
    pattern = gsub(pattern, "{.-}", [[([\w-.~%%!$&'()*+,;=@:]+)]])
    pattern = gsub(pattern, "%.", "\\.")

    return pattern
  end

  local function global(pattern)
    -- as per the RFC: https://tools.ietf.org/html/rfc3986#section-3.3
    local wildcard_regex = [[([\w-.~%%!$&'()*+,;=@:]+)]] -- using long Lua brackets [[...]], escaping `%`
    return pattern:gsub('?.*', ''):gsub("{.-}", wildcard_regex):gsub("%.", "\\.")
  end

  local function pcre(pattern)
    pattern = re_gsub(pattern, [[\?.*]], '')
    pattern = re_gsub(pattern, [[\{.+?\}]], [[([\w-.~%!$$&'()*+,;=@:]+)]])
    pattern = re_gsub(pattern, [[\.]], [[\.]])

    return pattern
  end

  local function pcre_jit(pattern)
    pattern = re_gsub(pattern, [[\?.*]], '', 'oj')
    pattern = re_gsub(pattern, [[\{.+?\}]], [[([\w-.~%!$$&'()*+,;=@:]+)]], 'oj')
    pattern = re_gsub(pattern, [[\.]], [[\.]], 'oj')

    return pattern
  end

  local function ffi_pcre(pattern)
    pattern = ffi_re_gsub(pattern, [[\?.*]], '')
    pattern = ffi_re_gsub(pattern, [[\{.+?\}]], [[([\w-.~%!$$&'()*+,;=@:]+)]])
    pattern = ffi_re_gsub(pattern, [[\.]], [[\.]])

    return pattern
  end

  local function ffi_pcre_jit(pattern)
    pattern = ffi_re_gsub(pattern, [[\?.*]], '', 'oj')
    pattern = ffi_re_gsub(pattern, [[\{.+?\}]], [[([\w-.~%!$$&'()*+,;=@:]+)]], 'oj')
    pattern = ffi_re_gsub(pattern, [[\.]], [[\.]], 'oj')

    return pattern
  end

  local str = '/tfgoodosod/{foo}/sdsd?asdasdd'

  print('loop:   ' , loop(str))
  print('closure:' , closure(str))
  print('global: ' , global(str))
  print('pcre:   ' , pcre(str))

  b:report('loop', function() return loop(str) end)
  b:report('gsub', function() return closure(str) end)
  b:report('string.gsub', function() return global(str) end)
  b:report('re.gsub', function() return pcre(str) end)
  b:report('re.gsub jit', function() return pcre_jit(str) end)

  b:report('ffi re.gsub', function() return ffi_pcre(str) end)
  b:report('ffi re.gsub jit', function() return ffi_pcre_jit(str) end)

  b:compare()
end)
</p>
```
```
Warming up --------------------------------------
                loop
     68527 i/100ms

                gsub
     69216 i/100ms

         string.gsub
     37823 i/100ms

             re.gsub
       221 i/100ms

         re.gsub jit
       796 i/100ms

         ffi re.gsub
     21117 i/100ms

     ffi re.gsub jit
     69048 i/100ms

Calculating -------------------------------------
   699969.0 (±2.3%) i/s -    3563404 in   5.093490s
   724349.3 (±1.3%) i/s -    3668448 in   5.065308s
   725619.1 (±1.9%) i/s -    3631008 in   5.005926s
      850.1 (±25.8%) i/s -       4199 in   5.331490s
     3794.1 (±11.9%) i/s -      19104 in   5.115673s
   293989.6 (±6.1%) i/s -    1478190 in   5.048416s
  2822971.0 (±11.0%) i/s -   13809600 in   5.023147s
```
So probably best to use the PCRE jited version :)

Reply:
Results of the same benchmark with JIT off:

```
Warming up --------------------------------------
                loop
     33967 i/100ms

                gsub
     33434 i/100ms

         string.gsub
     32714 i/100ms

             re.gsub
       205 i/100ms

         re.gsub jit
       596 i/100ms

         ffi re.gsub
      4601 i/100ms

     ffi re.gsub jit
      7278 i/100ms

Calculating -------------------------------------
   554070.5 (±2.5%) i/s -    2785294 in   5.030156s
   570121.9 (±2.5%) i/s -    2875324 in   5.046784s
   547568.0 (±1.7%) i/s -    2747976 in   5.020110s
      543.7 (±25.6%) i/s -       2665 in   5.228329s
     2492.8 (±4.6%) i/s -      12516 in   5.032186s
    48904.9 (±2.6%) i/s -     248454 in   5.083841s
    81249.3 (±4.0%) i/s -     407568 in   5.023757s
```